File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        cur = dummy
        carry = 0
        res = []
        while l1 or l2 or carry:
            
            num = 0
            if l1:
                num += l1.val
                l1 = l1.next

            if l2:
                num += l2.val
                l2 = l2.next
            num += carry 
               
            cur.next = ListNode(num%10)
            cur = cur.next
            carry = 1 if num>= 10 else 0

        return dummy.next

File: 0002-add-two-numbers/0002-add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode* head = (struct ListNode*)malloc(sizeof(struct ListNode));
    head->next = NULL;
    struct ListNode* cur = head;
    int carry = 0;
    int sum;

    while (l1 || l2 || carry){
        sum = 0;
        if (l1){
            sum += l1->val;
            l1 = l1->next;
        }

        if(l2){
            sum += l2->val;
            l2 = l2->next;
        }

        sum += carry;

        carry = sum/10;
        sum = sum%10;

        struct ListNode* new_node = (struct ListNode*)malloc(sizeof(struct ListNode));
        new_node -> next = NULL;
        new_node -> val = sum;

        cur->next = new_node;
        cur = cur->next;
    }

    return head->next;
}